server/src/main/java/io/crate/user/metadata/RolesMetadata.java

```diff
@@ -0,0 +1,233 @@
+                    boolean isUser = false;
+                    SecureHash secureHash = null;
+                    if (parser.nextToken() == XContentParser.Token.START_OBJECT) {
+                        while (secureHash == null && parser.nextToken() == XContentParser.Token.FIELD_NAME) {
+                            switch (parser.currentName()) {
+                                case "is_user":
+                                    parser.nextToken();
+                                    isUser = parser.booleanValue();
+                                    break;
+                                case "secure_hash":
+                                    secureHash = SecureHash.fromXContent(parser);
+                                    break;
+                                default:
+                                    throw new ElasticsearchParseException("failed to parse roles, unexpected token: " +
+                                        parser.currentToken());
+                            }
+                        }
+                        if (isUser) {
+                            roles.put(roleName, User.of(roleName, secureHash));
+                        } else {
+                            roles.put(roleName, Role.of(roleName, secureHash));
+                        }
+                    } else {
+                        // each custom metadata is packed inside an object.
+                        throw new ElasticsearchParseException("failed to parse roles, expected an object token at start");
+                    }
+                }
+            } else {
+                // each custom metadata is packed inside an object.
+                throw new ElasticsearchParseException("failed to parse roles, expected an object token at start");
+            }
+            if (parser.nextToken() != XContentParser.Token.END_OBJECT) {
+                // each custom metadata is packed inside an object.
+                // each custom must move the parser to the end otherwise possible following customs won't be read
+                throw new ElasticsearchParseException("failed to parse roles, expected an object token at the end");
+            }
+        }
+        return new RolesMetadata(roles);
+    }
+
+
+    @Override
+    public EnumSet<Metadata.XContentContext> context() {
+        return EnumSet.of(Metadata.XContentContext.GATEWAY, Metadata.XContentContext.SNAPSHOT);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        RolesMetadata that = (RolesMetadata) o;
+        return roles.equals(that.roles);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(roles);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return TYPE;
+    }
+
+    @Override
+    public Version getMinimalSupportedVersion() {
+        return Version.V_3_0_1;
```

Comment:
Shouldn't this be the current version (5.6)?

Reply:
of course, sorry, leftover